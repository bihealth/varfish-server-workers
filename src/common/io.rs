//! Common, IO-related code.

use std::{
    fs::File,
    io::{BufRead, BufReader, BufWriter, Write},
    path::Path,
};

use flate2::{bufread::MultiGzDecoder, write::GzEncoder, Compression};

/// Transparently open a file with gzip decoder.
pub fn open_read_maybe_gz<P>(path: P) -> Result<Box<dyn BufRead>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for reading", path.as_ref());
        let file = File::open(path)?;
        let bufreader = BufReader::new(file);
        let decoder = MultiGzDecoder::new(bufreader);
        Ok(Box::new(BufReader::new(decoder)))
    } else {
        tracing::trace!("Opening {:?} as plain text for reading", path.as_ref());
        let file = File::open(path).map(BufReader::new)?;
        Ok(Box::new(BufReader::new(file)))
    }
}

/// Transparently open afile with gzip encoder.
pub fn open_write_maybe_gz<P>(path: P) -> Result<Box<dyn Write>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for writing", path.as_ref());
        let file = File::create(path)?;
        let bufwriter = BufWriter::new(file);
        let encoder = GzEncoder::new(bufwriter, Compression::default());
        Ok(Box::new(encoder))
    } else {
        tracing::trace!("Opening {:?} as plain text for writing", path.as_ref());
        let file = File::create(path)?;
        Ok(Box::new(file))
    }
}

// The output is wrapped in a Result to allow matching on errors
// Returns an Iterator to the Reader of the lines of the file.
pub fn read_lines<P: AsRef<Path>>(
    filename: P,
) -> std::io::Result<std::io::Lines<std::io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(std::io::BufReader::new(file).lines())
}

#[cfg(test)]
mod test {
    use std::io::Read;

    #[rstest::rstest]
    #[case(true)]
    #[case(false)]
    fn open_write_maybe_gz(#[case] is_gzip: bool) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{:?}", is_gzip);

        let filename = if is_gzip { "test.txt" } else { "test.txt.gz" };
        let tmp_dir = temp_testdir::TempDir::default();

        {
            let mut f = super::open_write_maybe_gz(tmp_dir.join(filename))?;
            f.flush()?;
        }

        let mut f = std::fs::File::open(tmp_dir.join(filename)).map(std::io::BufReader::new)?;
        let mut buf = Vec::new();
        f.read_to_end(&mut buf)?;

        insta::assert_snapshot!(format!("{:x?}", &buf));

        Ok(())
    }

    #[test]
    fn read_lines() -> Result<(), anyhow::Error> {
        let lines = super::read_lines("tests/common/lines.txt")?.collect::<Result<Vec<_>, _>>()?;

        insta::assert_yaml_snapshot!(lines);

        Ok(())
    }
}
