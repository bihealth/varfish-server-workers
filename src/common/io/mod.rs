//! Common, IO-related code.

pub mod std;
