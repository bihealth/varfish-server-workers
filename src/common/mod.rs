//! Common functionality.

use std::ops::Range;

use biocommons_bioutils::assemblies::Assembly;
use byte_unit::Byte;
use clap::Parser;
use clap_verbosity_flag::{InfoLevel, Verbosity};
use indexmap::IndexMap;
use noodles_vcf as vcf;

pub mod noodles;
pub mod s3;

/// Commonly used command line arguments.
#[derive(Parser, Debug)]
pub struct Args {
    /// Verbosity of the program
    #[clap(flatten)]
    pub verbose: Verbosity<InfoLevel>,
}

impl Default for Args {
    fn default() -> Self {
        Self {
            verbose: Verbosity::new(0, 0),
        }
    }
}

/// Helper to print the current memory resident set size via `tracing`.
pub fn trace_rss_now() {
    let me = procfs::process::Process::myself().unwrap();
    let page_size = procfs::page_size();
    tracing::debug!(
        "RSS now: {}",
        Byte::from_bytes((me.stat().unwrap().rss * page_size) as u128).get_appropriate_unit(true)
    );
}

/// Definition of canonical chromosome names.
pub const CHROMS: &[&str] = &[
    "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17",
    "18", "19", "20", "21", "22", "X", "Y", "M",
];

/// Build mapping of chromosome names to chromosome counts.
pub fn build_chrom_map() -> IndexMap<String, usize> {
    let mut result = IndexMap::new();
    for (i, &chrom_name) in CHROMS.iter().enumerate() {
        result.insert(chrom_name.to_owned(), i);
        result.insert(format!("chr{chrom_name}").to_owned(), i);
    }
    result.insert("x".to_owned(), 22);
    result.insert("y".to_owned(), 23);
    result.insert("chrx".to_owned(), 22);
    result.insert("chry".to_owned(), 23);
    result.insert("mt".to_owned(), 24);
    result.insert("m".to_owned(), 24);
    result.insert("chrmt".to_owned(), 24);
    result.insert("chrm".to_owned(), 24);
    result.insert("MT".to_owned(), 24);
    result.insert("chrMT".to_owned(), 24);
    result
}

// Compute reciprocal overlap between two ranges.
pub fn reciprocal_overlap(lhs: Range<i32>, rhs: Range<i32>) -> f32 {
    let lhs_b = lhs.start;
    let lhs_e = lhs.end;
    let rhs_b = rhs.start;
    let rhs_e = rhs.end;
    let ovl_b = std::cmp::max(lhs_b, rhs_b);
    let ovl_e = std::cmp::min(lhs_e, rhs_e);
    if ovl_b >= ovl_e {
        0f32
    } else {
        let ovl_len = (ovl_e - ovl_b) as f32;
        let x1 = ovl_len / (lhs_e - lhs_b) as f32;
        let x2 = ovl_len / (rhs_e - rhs_b) as f32;
        x1.min(x2)
    }
}

/// Helper to convert ENSEMBL and RefSeq gene ID to u32.
pub fn numeric_gene_id(raw_id: &str) -> Result<u32, anyhow::Error> {
    let clean_id = if raw_id.starts_with("ENSG") {
        // Strip "ENSG" prefix and as many zeroes as follow
        raw_id
            .chars()
            .skip("ENSG".len())
            .skip_while(|c| *c == '0')
            .collect()
    } else {
        raw_id.to_owned()
    };

    clean_id
        .parse::<u32>()
        .map_err(|e| anyhow::anyhow!("could not parse gene id {:?}: {}", &clean_id, &e))
}

/// Select the genome release to use.
#[derive(
    clap::ValueEnum,
    Clone,
    Copy,
    Debug,
    strum::Display,
    PartialEq,
    Eq,
    enum_map::Enum,
    PartialOrd,
    Ord,
    Hash,
)]
pub enum GenomeRelease {
    // GRCh37 / hg19
    #[strum(serialize = "grch37")]
    Grch37,
    /// GRCh38 / hg38
    #[strum(serialize = "grch38")]
    Grch38,
}

impl GenomeRelease {
    pub fn name(&self) -> String {
        match self {
            GenomeRelease::Grch37 => String::from("GRCh37"),
            GenomeRelease::Grch38 => String::from("GRCh38"),
        }
    }
}

impl From<GenomeRelease> for Assembly {
    fn from(val: GenomeRelease) -> Self {
        match val {
            GenomeRelease::Grch37 => Assembly::Grch37p10,
            GenomeRelease::Grch38 => Assembly::Grch38,
        }
    }
}

impl From<Assembly> for GenomeRelease {
    fn from(assembly: Assembly) -> Self {
        match assembly {
            Assembly::Grch37 | Assembly::Grch37p10 => GenomeRelease::Grch37,
            Assembly::Grch38 => GenomeRelease::Grch38,
        }
    }
}

impl std::str::FromStr for GenomeRelease {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.to_ascii_lowercase();
        if s.starts_with("grch37") {
            Ok(GenomeRelease::Grch37)
        } else if s.starts_with("grch38") {
            Ok(GenomeRelease::Grch38)
        } else {
            Err(anyhow::anyhow!("Unknown genome release: {}", s))
        }
    }
}

/// Helper type for encoding genotypes in parsing.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Genotype {
    /// hom. ref.
    HomRef,
    /// het.
    Het,
    /// hom. alt.
    HomAlt,
    /// other, includes no-call
    WithNoCall,
}

impl std::str::FromStr for Genotype {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match s {
            "0/0" | "0|0" | "0" => Genotype::HomRef,
            "0/1" | "1/0" | "0|1" | "1|0" => Genotype::Het,
            "1/1" | "1|1" | "1" => Genotype::HomAlt,
            "./." | "./0" | "./1" | "0/." | "1/." => Genotype::WithNoCall,
            _ => anyhow::bail!("invalid genotype value: {:?}", s),
        })
    }
}

#[derive(Copy, Clone)]
pub enum Chrom {
    Auto, // or chrMT, but does not matter for carrier computation
    X,
    Y,
}

impl std::str::FromStr for Chrom {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match s {
            "X" => Chrom::X,
            "Y" => Chrom::Y,
            _ => Chrom::Auto,
        })
    }
}

/// The version of `varfish-server-worker` package.
pub const VERSION: &str = env!("CARGO_PKG_VERSION");

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn numeric_gene_id_simple() -> Result<(), anyhow::Error> {
        assert_eq!(1, numeric_gene_id("ENSG0000000001")?);
        assert_eq!(1, numeric_gene_id("ENSG1")?);
        assert_eq!(1, numeric_gene_id("1")?);

        Ok(())
    }
}

/// Return the version of the `varfish-server-worker` crate and `x.y.z` in tests.
pub fn worker_version() -> &'static str {
    if cfg!(test) {
        "x.y.z"
    } else {
        env!("CARGO_PKG_VERSION")
    }
}

/// Add contigs for GRCh37.
pub fn add_contigs_37(
    builder: vcf::header::Builder,
) -> Result<vcf::header::Builder, anyhow::Error> {
    use vcf::header::record::value::map::Contig;
    use vcf::header::record::value::Map;

    let mut builder = builder;

    let specs: &[(&str, usize); 25] = &[
        ("1", 249250621),
        ("2", 243199373),
        ("3", 198022430),
        ("4", 191154276),
        ("5", 180915260),
        ("6", 171115067),
        ("7", 159138663),
        ("8", 146364022),
        ("9", 141213431),
        ("10", 135534747),
        ("11", 135006516),
        ("12", 133851895),
        ("13", 115169878),
        ("14", 107349540),
        ("15", 102531392),
        ("16", 90354753),
        ("17", 81195210),
        ("18", 78077248),
        ("19", 59128983),
        ("20", 63025520),
        ("21", 48129895),
        ("22", 51304566),
        ("X", 155270560),
        ("Y", 59373566),
        ("MT", 16569),
    ];

    for (contig, length) in specs {
        builder = builder.add_contig(
            contig
                .parse()
                .map_err(|_| anyhow::anyhow!("invalid contig: {}", contig))?,
            Map::<Contig>::builder()
                .set_length(*length)
                .insert(
                    "assembly"
                        .parse()
                        .map_err(|_| anyhow::anyhow!("invalid key: assembly"))?,
                    "GRCh37",
                )
                .insert(
                    "species"
                        .parse()
                        .map_err(|_| anyhow::anyhow!("invalid key: species"))?,
                    "Homo sapiens",
                )
                .build()?,
        );
    }

    builder = builder.insert(
        "x-varfish-genome-build".parse()?,
        vcf::header::record::Value::from("GRCh37"),
    )?;

    Ok(builder)
}

/// Extract a PedigreeByName from the VCF header.
pub fn extract_pedigree_and_case_uuid(
    header: &vcf::Header,
) -> Result<(mehari::ped::PedigreeByName, uuid::Uuid), anyhow::Error> {
    let mut case_uuid = uuid::Uuid::nil();
    let mut pedigree = mehari::ped::PedigreeByName::default();

    if let vcf::header::record::value::Collection::Structured(sample_map) = header
        .other_records()
        .get("SAMPLE")
        .ok_or_else(|| anyhow::anyhow!("no SAMPLE record in VCF header"))?
    {
        for (sample_name, sample_values) in sample_map.iter() {
            let sex_value = sample_values
                .other_fields()
                .get("Sex")
                .ok_or_else(|| anyhow::anyhow!("no Sex field in SAMPLE header?"))?
                .as_ref();
            let sex_value = match sex_value {
                "Male" => mehari::ped::Sex::Male,
                "Female" => mehari::ped::Sex::Female,
                "Unknown" => mehari::ped::Sex::Unknown,
                _ => anyhow::bail!("invalid value for Sex: {}", sex_value),
            };

            let disease_value = sample_values
                .other_fields()
                .get("Disease")
                .ok_or_else(|| anyhow::anyhow!("no Disease field in SAMPLE header?"))?
                .as_ref();
            let disease_value = match disease_value {
                "Affected" => mehari::ped::Disease::Affected,
                "Unaffected" => mehari::ped::Disease::Unaffected,
                "Unknown" => mehari::ped::Disease::Unknown,
                _ => anyhow::bail!("invalid value for Disease: {}", disease_value),
            };

            pedigree.individuals.insert(
                sample_name.clone(),
                mehari::ped::Individual {
                    family: "FAM".into(),
                    name: sample_name.clone(),
                    sex: sex_value,
                    disease: disease_value,
                    ..Default::default()
                },
            );
        }
    }

    if let vcf::header::record::value::Collection::Structured(sample_map) = header
        .other_records()
        .get("PEDIGREE")
        .ok_or_else(|| anyhow::anyhow!("no PEDIGREE record in VCF header"))?
    {
        for (sample_name, pedigree_values) in sample_map.iter() {
            let father_value = pedigree_values.other_fields().get("Father");
            let mother_value = pedigree_values.other_fields().get("Mother");

            let individual = pedigree.individuals.get_mut(sample_name).ok_or_else(|| {
                anyhow::anyhow!("individual {} not found in SAMPLE header", sample_name)
            })?;
            if let Some(father_value) = father_value {
                individual.father = Some(father_value.clone());
            }
            if let Some(mother_value) = mother_value {
                individual.mother = Some(mother_value.clone());
            }
        }
    }

    if let vcf::header::record::value::Collection::Unstructured(lines) = header
        .other_records()
        .get("x-varfish-case-uuid")
        .ok_or_else(|| anyhow::anyhow!("no x-varfish-case-uuid record in VCF header"))?
    {
        case_uuid = lines
            .first()
            .ok_or_else(|| {
                anyhow::anyhow!("no x-varfish-case-uuid record in VCF header, but expected one")
            })?
            .parse()
            .map_err(|e| {
                anyhow::anyhow!(
                    "could not parse x-varfish-case-uuid record in VCF header: {}",
                    e
                )
            })?;
    }

    Ok((pedigree, case_uuid))
}

/// Add contigs for GRCh38.
pub fn add_contigs_38(
    builder: vcf::header::Builder,
) -> Result<vcf::header::Builder, anyhow::Error> {
    use vcf::header::record::value::map::Contig;
    use vcf::header::record::value::Map;

    let mut builder = builder;

    let specs: &[(&str, usize); 25] = &[
        ("chr1", 248956422),
        ("chr2", 242193529),
        ("chr3", 198295559),
        ("chr4", 190214555),
        ("chr5", 181538259),
        ("chr6", 170805979),
        ("chr7", 159345973),
        ("chr8", 145138636),
        ("chr9", 138394717),
        ("chr10", 133797422),
        ("chr11", 135086622),
        ("chr12", 133275309),
        ("chr13", 114364328),
        ("chr14", 107043718),
        ("chr15", 101991189),
        ("chr16", 90338345),
        ("chr17", 83257441),
        ("chr18", 80373285),
        ("chr19", 58617616),
        ("chr20", 64444167),
        ("chr21", 46709983),
        ("chr22", 50818468),
        ("chrX", 156040895),
        ("chrY", 57227415),
        ("chrM", 16569),
    ];

    for (contig, length) in specs {
        builder = builder.add_contig(
            contig
                .parse()
                .map_err(|_| anyhow::anyhow!("invalid contig: {}", contig))?,
            Map::<Contig>::builder()
                .set_length(*length)
                .insert(
                    "assembly"
                        .parse()
                        .map_err(|_| anyhow::anyhow!("invalid key: assembly"))?,
                    "GRCh38",
                )
                .insert(
                    "species"
                        .parse()
                        .map_err(|_| anyhow::anyhow!("invalid key: species"))?,
                    "Homo sapiens",
                )
                .build()?,
        );
    }

    builder = builder.insert(
        "x-varfish-genome-build".parse()?,
        vcf::header::record::Value::from("GRCh38"),
    )?;

    Ok(builder)
}

#[cfg(test)]
pub(crate) fn read_to_bytes<P>(path: P) -> Result<Vec<u8>, anyhow::Error>
where
    P: AsRef<std::path::Path>,
{
    use std::io::Read;

    let mut f = std::fs::File::open(&path).expect("no file found");
    let metadata = std::fs::metadata(&path).expect("unable to read metadata");
    let mut buffer = vec![0; metadata.len() as usize];
    f.read_exact(&mut buffer).expect("buffer overflow");
    Ok(buffer)
}

/// Enum for the supported TADs.
#[derive(
    serde::Serialize,
    serde::Deserialize,
    enum_map::Enum,
    PartialEq,
    Eq,
    Clone,
    Copy,
    Debug,
    Default,
    strum::EnumString,
    strum::Display,
)]
#[serde(rename_all = "lowercase")]
#[strum(serialize_all = "lowercase")]
pub enum TadSet {
    /// hESC
    #[default]
    Hesc,
}

#[macro_export]
macro_rules! flush_and_shutdown {
    ($output_writer:expr) => {
        let mut write = $output_writer.into_inner();
        // Flush all buffers.
        write
            .as_mut()
            .flush()
            .await
            .map_err(|e| anyhow::anyhow!("problem flushing output VCF file: {}", e))?;
        // Shutdown the writer.
        write
            .shutdown()
            .await
            .map_err(|e| anyhow::anyhow!("problem shutting down output VCF file: {}", e))?;
        // Give the OS some time to actually close the file.
        tokio::time::sleep(std::time::Duration::from_millis(200)).await;
    };
}

#[cfg(test)]
mod test {
    use noodles_vcf as vcf;

    #[test]
    fn trace_rss_now_smoke() {
        super::trace_rss_now();
    }

    #[test]
    fn build_chrom_map_snapshot() {
        let map = super::build_chrom_map();
        insta::assert_yaml_snapshot!(map);
    }

    #[rstest::rstest]
    #[case(0..10, 0..10, 1.0)]
    #[case(0..10, 5..15, 0.5)]
    #[case(5..15, 0..10, 0.5)]
    #[case(0..10, 10..20, 0.0)]
    #[case(0..2, 0..10, 0.2)]
    #[case(0..10, 0..2, 0.2)]
    fn reciprocal_overlap(
        #[case] lhs: std::ops::Range<i32>,
        #[case] rhs: std::ops::Range<i32>,
        #[case] expected: f32,
    ) {
        let actual = super::reciprocal_overlap(lhs, rhs);
        assert!(float_cmp::approx_eq!(f32, expected, actual, ulps = 2))
    }

    #[rstest::rstest]
    #[case("ENSG0000000142", 142)]
    #[case("42", 42)]
    fn numeric_gene_id(#[case] raw_id: &str, #[case] expected: u32) {
        let actual = super::numeric_gene_id(raw_id).unwrap();
        assert_eq!(expected, actual);
    }

    #[rstest::rstest]
    #[case(crate::common::GenomeRelease::Grch37, "GRCh37")]
    #[case(crate::common::GenomeRelease::Grch38, "GRCh38")]
    fn genome_release_name(#[case] release: super::GenomeRelease, #[case] expected: &str) {
        assert_eq!(expected, release.name());
    }

    #[rstest::rstest]
    #[case(
        crate::common::GenomeRelease::Grch37,
        biocommons_bioutils::assemblies::Assembly::Grch37p10
    )]
    #[case(
        crate::common::GenomeRelease::Grch38,
        biocommons_bioutils::assemblies::Assembly::Grch38
    )]
    fn assembly_from_genome_release(
        #[case] release: super::GenomeRelease,
        #[case] assembly: biocommons_bioutils::assemblies::Assembly,
    ) -> Result<(), anyhow::Error> {
        let res: biocommons_bioutils::assemblies::Assembly = release.into();

        assert_eq!(res, assembly);

        Ok(())
    }

    #[rstest::rstest]
    #[case(
        crate::common::GenomeRelease::Grch37,
        biocommons_bioutils::assemblies::Assembly::Grch37
    )]
    #[case(
        crate::common::GenomeRelease::Grch37,
        biocommons_bioutils::assemblies::Assembly::Grch37p10
    )]
    #[case(
        crate::common::GenomeRelease::Grch38,
        biocommons_bioutils::assemblies::Assembly::Grch38
    )]
    fn genome_release_from_assembly(
        #[case] release: super::GenomeRelease,
        #[case] assembly: biocommons_bioutils::assemblies::Assembly,
    ) -> Result<(), anyhow::Error> {
        let res: super::GenomeRelease = assembly.into();

        assert_eq!(res, release);

        Ok(())
    }

    #[rstest::rstest]
    #[case(crate::common::GenomeRelease::Grch37, "grch37")]
    #[case(crate::common::GenomeRelease::Grch38, "grch38")]
    fn genome_relese_from_str(
        #[case] release: super::GenomeRelease,
        #[case] s: &str,
    ) -> Result<(), anyhow::Error> {
        let res: super::GenomeRelease = s.parse()?;

        assert_eq!(res, release);

        Ok(())
    }

    #[test]
    fn extract_pedigree_snapshot() {
        let path = "tests/seqvars/aggregate/ingest.vcf";
        let mut vcf_reader = vcf::reader::Builder.build_from_path(path).unwrap();
        let header = vcf_reader.read_header().unwrap();

        let (pedigree, case_uuid) = super::extract_pedigree_and_case_uuid(&header).unwrap();
        insta::assert_debug_snapshot!(pedigree);
        insta::assert_debug_snapshot!(case_uuid);
    }
}
