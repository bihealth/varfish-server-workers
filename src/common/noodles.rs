//! Helper code for using noodles with S3.
//!
//! At the moment, we only support reading but not yet writing.
//!
//! The main reason is that the awslabs SDK for Rust currently supports easily
//! getting an `AsyncRead` from an S3 object but not an `AsyncWrite`.  For
//! this, we would have to create a wrapper that writes to a multipart upload
//! or similar.

use async_compression::tokio::bufread::GzipDecoder;
use mehari::common::{
    io::{std::is_gz, tokio::open_read_maybe_gz},
    noodles::AsyncVcfReader,
};
use noodles_bgzf as bgzf;
use noodles_core::Position;
use noodles_csi::{self as csi, index::reference_sequence::bin::Chunk};
use noodles_tabix as tabix;
use noodles_vcf as vcf;
use std::{path::Path, pin::Pin};
use tokio::io::{AsyncBufRead, BufReader};

/// Build TBI for file at `path_src` and write to `path_dst`.
pub async fn build_tbi<S, D>(path_src: S, path_dst: D) -> Result<(), anyhow::Error>
where
    S: AsRef<std::path::Path>,
    D: AsRef<std::path::Path>,
{
    let mut reader = tokio::fs::File::open(path_src.as_ref())
        .await
        .map(bgzf::AsyncReader::new)
        .map(vcf::AsyncReader::new)
        .map_err(|e| anyhow::anyhow!("error input file for tbi creation: {}", e))?;

    let header = reader
        .read_header()
        .await
        .map_err(|e| anyhow::anyhow!("error reading header: {}", e))?;

    let mut record = vcf::Record::default();

    let mut indexer = tabix::index::Indexer::default();
    indexer.set_header(csi::index::header::Builder::vcf().build());

    let mut start_position = reader.get_ref().virtual_position();

    while reader
        .read_record(&header, &mut record)
        .await
        .map_err(|e| anyhow::anyhow!("problem reading record: {}", e))?
        != 0
    {
        let end_position = reader.get_ref().virtual_position();
        let chunk = Chunk::new(start_position, end_position);

        let reference_sequence_name = record.chromosome().to_string();
        let start = Position::try_from(usize::from(record.position()))
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
            .map_err(|e| anyhow::anyhow!("error converting start position: {}", e))?;
        let end = record
            .end()
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
            .and_then(|position| {
                Position::try_from(usize::from(position))
                    .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
            })
            .map_err(|e| anyhow::anyhow!("error converting end position: {}", e))?;

        indexer
            .add_record(&reference_sequence_name, start, end, chunk)
            .map_err(|e| anyhow::anyhow!("error adding record to tabix index: {}", e))?;

        start_position = end_position;
    }

    let index = indexer.build();

    {
        let mut writer = tokio::fs::File::create(path_dst.as_ref())
            .await
            .map(tokio::io::BufWriter::new)
            .map(tabix::AsyncWriter::new)
            .map_err(|e| anyhow::anyhow!("error output file for tbi creation: {}", e))?;
        writer
            .write_index(&index)
            .await
            .map_err(|e| anyhow::anyhow!("error writing tabix index: {}", e))?;
        writer
            .shutdown()
            .await
            .map_err(|e| anyhow::anyhow!("error flushing tabix index: {}", e))?;

        tokio::time::sleep(std::time::Duration::from_millis(200)).await;
    }

    Ok(())
}

/// Open plain text or gzip reader via S3.
pub async fn s3_open_read_maybe_gz<P>(path: P) -> Result<Pin<Box<dyn AsyncBufRead>>, anyhow::Error>
where
    P: AsRef<Path>,
{
    let path_string = format!("{}", path.as_ref().display());
    let (bucket, key) = if let Some((bucket, key)) = path_string.split_once('/') {
        (bucket.to_string(), key.to_string())
    } else {
        anyhow::bail!("invalid S3 path: {}", path.as_ref().display());
    };

    let config = aws_config::load_from_env().await;
    let client = aws_sdk_s3::Client::new(&config);
    let object = client.get_object().bucket(&bucket).key(&key).send().await?;

    let path_is_gzip = is_gz(path.as_ref());
    tracing::trace!(
        "Opening S3 object {} as {} for reading (async)",
        path.as_ref().display(),
        if path_is_gzip {
            "gzip (allow multi-member)"
        } else {
            "plain text"
        }
    );

    if path_is_gzip {
        let bufreader = BufReader::new(object.body.into_async_read());
        let decoder = {
            let mut decoder = GzipDecoder::new(bufreader);
            decoder.multiple_members(true);
            decoder
        };
        Ok(Box::pin(BufReader::new(decoder)))
    } else {
        Ok(Box::pin(BufReader::new(object.body.into_async_read())))
    }
}

/// Helper function that opens a list of paths as VCF readers.
pub async fn open_vcf_readers(paths: &[String]) -> Result<Vec<AsyncVcfReader>, anyhow::Error> {
    let mut result = Vec::new();
    for path in paths.iter() {
        let buf_read = if super::s3::s3_mode() && !path.starts_with('/') {
            s3_open_read_maybe_gz(path).await?
        } else {
            open_read_maybe_gz(path).await?
        };
        result.push(AsyncVcfReader::new(buf_read));
    }
    Ok(result)
}

/// Helper function that opens one VCF reader at the given path.
///
/// The behaviour is as follows:
///
/// - If `path_in` is "-" then open stdin and read as plain text.
/// - If environment variable `AWS_PROFILE` is set to "varfish-s3" then enable S3 mode.
/// - If `path_in` is absolute or S3 mode is disabled then open `path_in` as local file
/// - Otherwise, attempt to open `path_in` as S3 object.
pub async fn open_vcf_reader(path_in: &str) -> Result<AsyncVcfReader, anyhow::Error> {
    let s3_mode = match std::env::var("AWS_PROFILE") {
        Ok(s) => s == "varfish-s3",
        _ => false,
    };
    if s3_mode && !path_in.starts_with('/') {
        Ok(vcf::AsyncReader::new(
            s3_open_read_maybe_gz(path_in)
                .await
                .map_err(|e| anyhow::anyhow!("could not build VCF reader from S3 file: {}", e))?,
        ))
    } else {
        Ok(vcf::AsyncReader::new(
            open_read_maybe_gz(path_in).await.map_err(|e| {
                anyhow::anyhow!("could not build VCF reader from local file: {}", e)
            })?,
        ))
    }
}

#[cfg(test)]
mod test {
    #[tokio::test]
    async fn build_tbi() -> Result<(), anyhow::Error> {
        let tmpdir = temp_testdir::TempDir::default();

        let path_in: String = "tests/seqvars/ingest/NA12878_dragen.vcf.gz".into();
        let path_out: String = tmpdir
            .join("out.vcf.gz.tbi")
            .to_str()
            .expect("invalid path")
            .into();
        super::build_tbi(&path_in, &path_out).await?;

        // This must be fixed first https://github.com/zaeleus/noodles/issues/213
        // let mut buffer: Vec<u8> = Vec::new();
        // hxdmp::hexdump(&crate::common::read_to_bytes(&path_out)?, &mut buffer)?;
        // insta::assert_snapshot!(String::from_utf8_lossy(&buffer));

        Ok(())
    }
}
