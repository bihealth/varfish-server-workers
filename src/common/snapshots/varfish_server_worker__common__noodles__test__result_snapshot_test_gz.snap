---
source: src/common/noodles.rs
expression: "String::from_utf8_lossy(&buffer)"
---
0000: 1F 8B 08 04 00 00 00 00 00 FF 06 00 42 43 02 00  ............BC..
0016: B8 00 ED DC BF 09 C2 40 14 07 E0 17 11 D1 D6 5A  .......@.......Z
0032: 8C 20 D8 6B 63 61 21 D8 45 10 9B D4 6E E0 08 2E  ...kca!.E...n...
0048: E1 06 2E E0 3C 8E 63 C2 DD 15 C9 0C DF 07 97 97  ....<.c.........
0064: 7B F7 6F 83 5F 7B 69 AA 49 44 F4 A3 CA B5 B7 CD  {.o._{i.ID......
0080: 75 D6 8D FD 31 6E 6D 4C BB BF CD 2A ED FA 3C D3  u...1nmL...*..<.
0096: EA FC 95 EA 3A F7 CF 75 9A 3F 4E A9 D6 B9 5F E6  ....:..u.?N..._.
0112: E5 DC F5 9D 5E 2A FB CB 3D F7 18 FA 2D 02 00 00  ....^*..=...-...
0128: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
0144: 00 06 C6 D9 77 25 EB AE CF B9 6B 96 29 07 AF E4  ....w%....k.)...
0160: DC 45 7C 77 FD B7 E4 E0 8D FB 87 D1 DD C3 B3 C9  .E|w............
0176: 1F 96 E4 1B E0 BA 4F 00 00                       ......O..
